Add basic handshake defragmentation tests in ssl-opt

minosgalanakis opened the pull request:
## Description
Implements a portion of the testing needed for #9872. This pr extends #9928 and resolves #9887



## PR checklist

Please remove the segment/s on either side of the | symbol as appropriate, and add any relevant link/s to the end of the line.
If the provided content is part of the present PR remove the # symbol.

- [x] **changelog** provided.
- [x] **development PR** provided
- [x] **TF-PSA-Crypto PR** not required because:  ssl-opt testing changes only
- [x] **framework PR** not required
- [x] **3.6 PR** provided #10016
- [x] **2.28 PR** not required because:  Not backported to 2.28
- **tests**  provided

mpg commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,344 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+# OpenSSL does not allow max_send_frag to be less than 512
+# so we use split_send_frag instead for tests lower than 512 below.
+
+# There is an issue with OpenSSL when fragmenting with values less
```
I think this comment is outdated since we're now using OpenSSL 3.x which is unaffected, and we do test with values lower than 512 with 1.2 below.

gilles-peskine-arm replied:
Indeed. Now the comment should be “despite what the manual says, -split_send_frag works out of the box with OpenSSL >=3.0”.

mpg replied:
The first two lines of the comment are still correct, but I was commenting on the last two lines ("There is an issue ...").

mpg commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,344 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            "$P_CLI debug_level=4 " \
+            0 \
+            -c "reassembled record" \
+            -c "handshake fragment: 0 \\.\\. 5 of [0-9]\\+ msglen 5" \
```
Minor: we should be able to go down to 4 (that's the minimum supported by our code, because it's the size of then handshake header).

gilles-peskine-arm replied:
Indeed. Please make sure we have positive test cases with `-split_send_frag 4`, and negative test cases with `-split_send_frag 3`.

mpg reviewed (commented):
(Posting partial early review.)

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,344 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
             -c "Handshake was completed" \
             -s "dumping .client hello, compression. (2 bytes)"
 
+# Handshake defragmentation testing
```
The default certificate depends on the configuration. I think `server5.crt` (signed with secp256r1, without multiple names or any lengthy extension) happens the smallest certificate we use for TLS, but this might not always be the case. So if we need a specific size, it might be better to list the certificate explicitly. That will prevent the test case from running in configurations without the required crypto, but I think that's ok here, since this test isn't related to what asymmetric crypto is used.

mpg replied:
> The default certificate depends on the configuration. 

Our default certificate, yes, but the certificate used by OpenSSL (ie, the party that fragments) is determined by its command line options, and `O_NEXT_SRV` has `server5.crt`. If we want to be super explicit though we could use `O_NEXT_SRV_NO_CERT` then append the certificate options manually, though I'm not sure we should really do that.

mpg replied:
The Certificate message that's going to be fragmented is always going to be the one sent by OpenSSL (we enable client auth so that it sends one) , and the certificate it contains is determined by ssl-opt, not `mbedtls_config.h`.

That said, point taken that "the default certificate" was a bad choice of words here, should have said "the certificate sent by OpenSSL".

In the interest of clarity, the comment should also mention that when OpenSSL is the server, then the certificate is defined as part of `O_NEXT_SRV` while when it's acting as a client, it's part of the test data. (And explain that we're using client auth for the purpose of ensuring at least one large message.)

minosgalanakis commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,344 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
             -s "dumping .client hello, compression. (2 bytes)"
 
+# Handshake defragmentation testing
+
```
I have chosen the test lines specifically that they only appear when fragmentation is happening.  There is no harm in adding the negative assertion tests though.

mpg replied:
> I have chosen the test lines specifically that they only appear when fragmentation is happening.

I know. The point of tests is to ensure that this choice was indeed correct.

I used to believe that "handshake fragment" only appeared when fragmentation was happening, I'm now quite convinced I was wrong and it does appear all the time, whether fragmentation happens or not.

I still believe that "reassembled record" and "waiting fro more fragments" only appear when fragmentation actually happens, but I could be wrong about that too. So, let's have tests.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -13873,6 +13873,16 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            "$P_CLI debug_level=4 " \
+            0 \
+            -C "reassembled record" \
+            -C "handshake fragment:" \
```
I believe this should be `-c`.

minosgalanakis replied:
Apparently the presence of this text really depends on configuration, so I am inclined to remove it.

I could try to be more strict using something like `handshake fragment: 0 \\.\\. [0-9] of [0-9]\\+ msglen [0-9]" \` but I am not certain it will add much.

mpg replied:
From reading the source code I really don't think it depends on the configuration (it's specific to TLS as opposed to DTLS, but TLS is always compiled in, and those tests are all TLS anyway).

I think it's just always present and the assertion here should be `-c`.

mpg replied:
If I'm right, this test should never pass and in all components it will be either SKIP, or FAIL with the reason that this message was present in the client log while the test is saying it should be absent.

If I'm wrong, please show me a configuration where the message is absent.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -13873,6 +13873,10 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+
+# To warrant that the handhake messages are large enough and need to be split
+# into fragments, the tests require certificate authentication, and will
+# use either the default per configuration certificate or the server5.crt (548 Bytes).
```
As discussed above, the party that fragments is always OpenSSL and it always uses server5.crt, either from `O_NEXT_SRV` or from the test data.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -14229,6 +14240,29 @@ run_test    "Handshake defragmentation on server: len=5, TLS 1.3" \
+
+requires_openssl_3_x
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
```
Please add a comment explaining this unusual requirement, otherwise future readers might think it was a mistake.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -14086,6 +14086,28 @@ run_test    "Handshake defragmentation on client: len=5, TLS 1.3" \
+requires_openssl_3_x
+requires_protocol_version tls13
+requires_certificate_authentication
+run_test    "Handshake defragmentation on client: len=3, TLS 1.3" \
```
Please document this in the ChangeLog entry: "The initial fragment for each handshake message must be at least 4 bytes."

mpg commented on tests/ssl-opt.sh:
```diff
@@ -14263,6 +14285,29 @@ run_test    "Handshake defragmentation on server: len=4, TLS 1.2" \
+            -s "<= parse client hello" \
+            -s "handshake message too short: 3" \
+            -s "SSL - An invalid SSL record was received"
+
```
Please also add a negative test where the server has `force_version=tls12` with a reasonable fragment size (anywhere between 4 and 40). We expect this to fail too, but please check what the message is.

Also, please document this in the ChangeLog entry: "Server-side, defragmentation of the ClientHello message is only supported if the server accepts TLS 1.3 (regardless of whether the ClientHello is 1.3 or 1.2). That is, servers configured (either at compile time or at runtime) to only accept TLS 1.2 will still fail the handshake if the ClientHello message is fragmented."

mpg commented on tests/ssl-opt.sh:
```diff
@@ -13899,9 +13899,13 @@ run_test    "Handshake defragmentation on client: len=512, TLS 1.3" \
             -c "handshake fragment: 0 \\.\\. 512 of [0-9]\\+ msglen 512" \
             -c "waiting for more fragments (512 of [0-9]\\+"
 
+# Since the removal of the DSA-RSA key exchange, the default openssl server
```
Typo: you mean DHE-RSA :)

minosgalanakis commented:
Quick update:

I have addressed all the comments, cleaned up the commit history to make it easier to review, and increased coverage. 

I will be sharing a table with what is covered for design review purposes.

mpg commented:
Now that #9872 is merged, can you rebase on top of it?

Also, can you add a commit disabling (by prepending `skip_next_test`) the failings tests (TLS 1.2 with < 16 fragments) temporarily, just to make it easier to check that those are the only failures (ie, the CI should come back green with those disabled).

Once we've agreed on a solution to the underlying problem, we'll re-enable those tests in some form (probably just assert failure when using 1.2 and with fragment length < 16, but we can't do that right now, as the handshake succeeds with ChachaPoly).

gilles-peskine-arm commented:
> Also, can you add a commit disabling (by prepending skip_next_test) the failings tests (TLS 1.2 with < 16 fragments) temporarily, just to make it easier to check that those are the only failures (ie, the CI should come back green with those disabled).

Reminder that you can check that easily in [failures.csv](https://mbedtls.trustedfirmware.org/job/mbed-tls-pr-head/job/PR-9989-head/13/artifact/failures.csv). (Except for failures that aren't unit tests or SSL tests, such as [build failures](https://github.com/Mbed-TLS/mbedtls-framework/pull/129)).

Neustradamus commented:
To follow this PR.

minosgalanakis commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,596 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+# of the fragmentation operations is OpenSSL and will always use server5.crt (548 Bytes)
+# either from O_NEXT_SRV or test data.
+requires_openssl_3_x
+requires_protocol_version tls13
```
> requires_protocol_version tls13 is not needed when the invocation of a program contains -tls1_3 or force_version=tls13. Same with 1.2.

Unfortunately that does not always seem to be case.

In the latest CI_Run `test_tls12_only` and `full_without_ecdhe_ecdsa_and_tls13` failed because `-tls1_3` tests were executed while they [should have  been](https://github.com/Mbed-TLS/mbedtls/blob/3e86e97d68aba600d29c8b9c8fe3431bd75371af/tests/ssl-opt.sh#L14129-L14136) skipped being tls 1.3 dependant.

I do not fully understand how the automatic detection work, which is why I added the `requires_protocol_version tls13`

I could add it specifically for the cases that fail if that would make sense?

gilles-peskine-arm replied:
Ah, I was mistaken earlier. We only automate the protocol dependency for `force_version=…`, not for `-tls1_3`.

Thinking about it, I suspect this was deliberate, so that we don't catch negative test cases where we test what happens if the peer wants a version that we don't support.

[This should definitely be better documented](https://github.com/Mbed-TLS/mbedtls/issues/10009)!

mpg commented:
The CI wants you do add three entries [there](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/scripts/analyze_outcomes.py#L52) for the 3 tests that are forcibly skipped in ssl-opt.sh, and then it should finally be fully happy!

minosgalanakis commented:
To faciliate the reiview and design review purposes I have compiled a table for the coverage.

| Client/Server | Fragment Length | TLS 1.2 | TLS 1.3 | Negative |
|--------------|----------------------|--------|--------|----------|
| Client      | No Fragmentation      |        | ✔      |          |
| Client      | 513                   |        | ✔      |          |
| Client      | 513                   | ✔      |        |          |
| Client      | 512                   |        | ✔      |          |
| Client      | 512                   | ✔      |        |          |
| Client      | 256                   |        | ✔      |          |
| Client      | 256                   | ✔      |        |          |
| Client      | 128                   |        | ✔      |          |
| Client      | 128                   | ✔      |        |          |
| Client      | 64                    |        | ✔      |          |
| Client      | 64                    | ✔      |        |          |
| Client      | 36                    |        | ✔      |          |
| Client      | 36                    | ✔      |        |          |
| Client      | 32                    |        | ✔      |          |
| Client      | 32                    | ✔      |        |          |
| Client      | 16                    |        | ✔      |          |
| Client      | 16                    | ✔      |        |          |
| Client      | 13                    |        | ✔      |          |
| Client      | 13                    | ✔      |        |          |
| Client      | 5                     |        | ✔      |          |
| Client      | 5                     | ✔      |        |          |
| Client      | 4                     |        | ✔      |          |
| Client      | 4                     | ✔      |        |          |
| Client      | 3                     |        | ✔      | ✔        |
| Client      | 3                     | ✔      |        | ✔        |
| Server      | No Fragmentation      |        | ✔      |          |
| Server      | 513                   |        | ✔      |          |
| Server      | 513                   | ✔      |        |          |
| Server      | 512                   |        | ✔      |          |
| Server      | 512                   | ✔      |        |          |
| Server      | 256                   |        | ✔      |          |
| Server      | 256                   | ✔      |        |          |
| Server      | 128                   |        | ✔      |          |
| Server      | 128                   | ✔      |        |          |
| Server      | 64                    |        | ✔      |          |
| Server      | 64                    | ✔      |        |          |
| Server      | 36                    |        | ✔      |          |
| Server      | 36                    | ✔      |        |          |
| Server      | 32                    |        | ✔      |          |
| Server      | 32                    | ✔      |        |          |
| Server      | 16                    |        | ✔      |          |
| Server      | 16                    | ✔      |        |          |
| Server      | 13                    |        | ✔      |          |
| Server      | 13                    | ✔      |        |          |
| Server      | 5                     |        | ✔      |          |
| Server      | 5                     | ✔      |        |          |
| Server      | 4                     |        | ✔      |          |
| Server      | 4                     | ✔      |        |          |
| Server      | 3                     |        | ✔      | ✔        |
| Server      | 3  (TLS 1.3 -> 1.2)  | ✔      |   ✔     | ✔        |
| Server      | 32 (TLS 1.3 -> 1.2)   | ✔      |   ✔     | ✔        |

mpg reviewed (commented):
LGTM. I have some minor feedback that can either be addressed here (while the 3.6 backport is being prepared) or left for a follow-up - no blocker so far IMO.

I'd still like to make another pass tomorrow before I formally approve, as the relatively high volume makes it hard to remain focused and easy to overlook things.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,546 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            -s "handshake message too short: 3" \
+            -s "SSL - An invalid SSL record was received"
+
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
+requires_certificate_authentication
+run_test    "Handshake defragmentation on server: len=32, TLS 1.3 Client-Hallo -> 1.2 Handhsake" \
+            "$P_SRV debug_level=4 force_version=tls12 auth_mode=required" \
```
Copypasta: this is testing a 1.2 ClientHello with the 1.2 parser.

```suggestion
requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
requires_certificate_authentication
run_test    "Handshake defragmentation on server: len=32, TLS 1.2 ClientHello" \
            "$P_SRV debug_level=4 force_version=tls12 auth_mode=required" \
```

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,546 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            1 \
+            -s "The SSL configuration is tls12 only" \
+            -s "bad client hello message" \
+            -s "SSL - A message could not be parsed due to a syntactic error"
```
Can wait for a follow-up: We should check `"ClientHello fragmentation not supported"` in the log.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,546 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            "$P_SRV debug_level=4 auth_mode=required" \
+            "$O_NEXT_CLI -cert $DATA_FILES_PATH/server5.crt -key $DATA_FILES_PATH/server5.key" \
+            0 \
+            -C "reassembled record" \
+            -C "waiting for more fragments"
```
Copypasta
```suggestion
            -S "reassembled record" \
            -S "waiting for more fragments"
```

gilles-peskine-arm reviewed (changes requested):
I've reviewed only for correctness, not for completeness, since we don't have a clear coverage objective here.

I'd like to fix the copypasta and the comment that would be wrong in 3.6. The rest can wait until follow-ups.

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,546 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
+requires_certificate_authentication
+run_test    "Handshake defragmentation on client: len=512, TLS 1.3" \
```
Generally speaking, I'm in favor of less repetitive code. Having nearly-identical blocks of ~10 lines has a high risk of mistakes and can obscure the forest for the trees.

Regarding the method, I'm not so keen on a shell function, but not fundamentally opposed. They have several pitfalls:

* The shell is limited as a programming language, and not very well known.
* It's harder to locate and figure out test cases that are generated on the fly.
* The function must not depend on the library configuration or the environment — it must always emit the same test cases as `ssl-opt.sh --list-test-cases`. For example, it must make `requires_xxx` calls and a final call to `run_test`, never return early.

There's another possible approach, which is to generate test cases from a Python script, with an intermediate file containing the individual test cases. It's a bit heavier weight to get going, but then has fewer pitfalls.

In terms of existing infrastructure, we currently have [a few functions (`run_test_psa`, `run_test_memory_after_handshake` and their variants)](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.6.2/tests/ssl-opt.sh#L1780) and one Python script [`generate_tls13_compat_tests.py`](https://github.com/Mbed-TLS/mbedtls-framework/blob/2eb60b8da3745f8882741f6fa4c2b316abf5478a/scripts/generate_tls13_compat_tests.py).

minosgalanakis commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,546 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+            -c "waiting for more fragments (512 of [0-9]\\+"
+
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3
+requires_certificate_authentication
```
Because since #9917 the supported openssl ciphers for TLS.1.2 and what our server provides, do not have any cipher in common when ECDSA is removed.

With TLS1.3 there are plenty of alternatives

gilles-peskine-arm replied:
Surely we have RSA-ECDHE cipher suites in common? You can configure the library with ECDH but not ECDSA. Maybe we don't run TLS tests in such a configuration from `all.sh`, but it is possible.

mpg replied:
> Surely we have RSA-ECDHE cipher suites in common?

Yes but not when the server only has an ECC cert.

I _think_ the reason we don't need to add an ECDSA requirement for 1.3 is that there the autodetection logic works: it sees one party using server5 and [requires ECDSA](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/ssl-opt.sh#L517).

I also think the reason the autodetection does not work for 1.2 is that there things are more complex: if both peers are Mbed TLS, then ECC cert doesn't necessarily mean ECDHE-ECDSA key exchange, a static ECDH key exchange could be used as well so we [require any key exchange that can use an ECC cert](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/ssl-opt.sh#L527).

Except OpenSSL doesn't allow static ECDH by default (or even doesn't support it at all, didn't check recently). So the autodetection logic [should notice](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/ssl-opt.sh#L1285) that on of the peers doesn't support static ECDH, and require ECDHE-ECDSA, but apparently it doesn't, so we find ourselves adding the requirement manually. (Or perhaps it's [the part that detects the version](https://github.com/Mbed-TLS/mbedtls/blob/development/tests/ssl-opt.sh#L1668) that's broken?)

I've not actually tested whether this reasoning is correct, but that's what I'm thinking so far based on reading the code. Anyway, from CI failures in past version of this PR, it would appear that the autodetection logic is working for the TLS 1.3 test cases but not for the 1.2 test cases in this series, whatever the reason.

gilles-peskine-arm reviewed (changes requested):
There's still one spurious dependency on `requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_3`: https://github.com/Mbed-TLS/mbedtls/pull/9989#discussion_r1973545194 . Other than that LGTM

gilles-peskine-arm commented on tests/ssl-opt.sh:
```diff
@@ -13872,6 +13872,541 @@ run_test    "TLS 1.2 ClientHello indicating support for deflate compression meth
+
+requires_config_enabled MBEDTLS_SSL_PROTO_TLS1_2
+requires_certificate_authentication
+requires_config_enabled MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED
```
I had missed this spurious dependency, good call for removing it from the backport: the negative test for handshake too small doesn't require ECDSA support. In fact it doesn't need certificate authentication at all. We can fix this later.

gilles-peskine-arm reviewed (approved):
LGTM

minosgalanakis commented:
One test failedin ` all_u16-test_tls13_only_psk`, because I removed the certifcate requirement for the negative testing. Re-introduced that and should be green.

mpg commented on ChangeLog.d/tls-hs-defrag-in.txt:
```diff
@@ -3,3 +3,10 @@ Bugfix
      version could be affected in principle, and both are fixed now).
+     The initial fragment for each handshake message must be at least 4 bytes.
+
+     Server-side, defragmentation of the ClientHello message is only
```
Good point. Generally speaking, we (and I'm as guilty of it as anyone) tend to put too much information in the ChangeLog that should really be in the documentation.

In this case though, I'm not entirely sure where this would fit in `ssl.h`. In the documentation of `ssl_handshake()`? 

Ideally, I think in addition to doxygen comments, we should have a (versioned) summary document for what's implemented or not or partially for TLS 1.2, DTLS and for X.509. I think we have something like this for TLS 1.3 (`docs/architecture/tls13-support.md`). For TLS 1.2 I thought there was something (as list of implemented RFCs, probably outdated) in the old Knowledge Base but I can't find it any more.

mpg commented on tests/ssl-opt.sh:
```diff
@@ -14220,12 +14217,12 @@ run_test    "Handshake defragmentation on server: len=128, TLS 1.3" \
             -s "waiting for more fragments (128"
 
-# Server-side ClientHello degfragmentation is only supported for MBEDTLS_SSL_PROTO_TLS1_3. For TLS 1.2 testing
+# Server-side ClientHello defragmentationis only supported for MBEDTLS_SSL_PROTO_TLS1_3. For TLS 1.2 testing
```
Introducing a type: missing space defragmentationis

mpg reviewed (approved):
LGTM, thanks!

gilles-peskine-arm reviewed (approved):
We'll need to have some tests around fragmentation that run with PSK, but ok for now.
